piker/brokers/kucoin.py

```diff
@@ -0,0 +1,685 @@
+    server every ping_interval so Kucoin doesn't drop
+    our connection
+
+    '''
+    async with trio.open_nursery() as n:
+        # TODO: cache this task so it's only called once
+        async def ping_server():
+            while True:
+                await trio.sleep((ping_interval - 1000) / 1000)
+                await ws.send_msg({'id': connect_id, 'type': 'ping'})
+
+        log.info(f'Starting ping task for kucoin ws connection')
+        n.start_soon(ping_server)
+
+        yield
+
+        n.cancel_scope.cancel()
+
+
+async def stream_quotes(
+    send_chan: trio.abc.SendChannel,
+    symbols: list[str],
+    feed_is_live: trio.Event,
+    loglevel: str = '',
+    # startup sync
+    task_status: TaskStatus[tuple[dict, dict]] = trio.TASK_STATUS_IGNORED,
+) -> None:
+    '''
+    Required piker api to stream real-time data.
+    Where the rubber hits the road baby
+
+    '''
+    async with open_cached_client('kucoin') as client:
+        token, ping_interval = await client._get_ws_token()
+        connect_id = str(uuid4())
+        pairs = await client.cache_pairs()
+
+    # open ping task
+        async with (
+            open_autorecon_ws(
+                f'wss://ws-api-spot.kucoin.com/?token={token}&[connectId={connect_id}]'
+            ) as ws,
+            open_ping_task(ws, ping_interval, connect_id),
+        ):
+            log.info('Starting up quote stream')
+            # loop through symbols and sub to feedz
+            for sym in symbols:
+                pair: KucoinMktPair = pairs[sym]
+                kucoin_sym = pair.symbol
+
+                init_msgs = {
+                    # pass back token, and bool, signalling if we're the writer
+                    # and that history has been written
+                    sym: {
+                        'symbol_info': {
+                            'asset_type': 'crypto',
+                            'price_tick_size': float(pair.baseIncrement),
+                            'lot_tick_size': float(pair.baseMinSize),
+                        },
+                        'shm_write_opts': {'sum_tick_vml': False},
+                        'fqsn': sym,
+                    }
+                }
+
+
+                async with (
+                    subscribe(ws, connect_id, kucoin_sym),
+                    stream_messages(ws, sym) as msg_gen,
+                ):
+                    typ, quote = await anext(msg_gen)
+                    while typ != 'trade':
+                        # take care to not unblock here until we get a real trade quote
+                        typ, quote = await anext(msg_gen)
+
+                    task_status.started((init_msgs, quote))
+                    feed_is_live.set()
+
+                    async for typ, msg in msg_gen:
+                        await send_chan.send({sym: msg})
+
+@acm
+async def subscribe(ws: wsproto.WSConnection, connect_id, sym):
+    # level 2 sub
+    await ws.send_msg({
```

Comment:
i'm surprised you can't do both of these subs in one msg?

Reply:
Seems like what's even the point the way their API works? https://docs.kucoin.com/#multiplex @goodboy